File: experiments/engine-kt/tools/src/main/kotlin/com/garive/eng/kt/tools/GovernedTypes.kt
package com.garive.eng.kt.tools

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** Structural construction result for C5 values. */
public sealed interface GovernedValueResult<out T> {
    /** Valid immutable value. */
    public data class Success<T>(public val value: T) : GovernedValueResult<T>
    /** Stable structural failure. */
    public data class Failure(public val label: String) : GovernedValueResult<Nothing>
}

private fun required(value: String, label: String): GovernedValueResult.Failure? =
    if (value.isEmpty()) GovernedValueResult.Failure(label) else null

/** Runtime-owned non-empty tool invocation identity. */
public class ToolInvocationId private constructor(public val value: String) {
    public companion object {
        /** Validates one identity. */
        public fun create(value: String): GovernedValueResult<ToolInvocationId> =
            required(value, "tool invocation") ?: GovernedValueResult.Success(ToolInvocationId(value))
    }
    public override fun equals(other: Any?): Boolean = other is ToolInvocationId && value == other.value
    public override fun hashCode(): Int = value.hashCode()
}

/** Runtime-owned non-empty interaction identity. */
public class InteractionId private constructor(public val value: String) {
    public companion object {
        /** Validates one identity. */
        public fun create(value: String): GovernedValueResult<InteractionId> =
            required(value, "interaction") ?: GovernedValueResult.Success(InteractionId(value))
    }
    public override fun equals(other: Any?): Boolean = other is InteractionId && value == other.value
    public override fun hashCode(): Int = value.hashCode()
}

/** Runtime-owned non-empty grant identity. */
public class GrantId private constructor(public val value: String) {
    public companion object {
        /** Validates one identity. */
        public fun create(value: String): GovernedValueResult<GrantId> =
            required(value, "grant") ?: GovernedValueResult.Success(GrantId(value))
    }
    public override fun equals(other: Any?): Boolean = other is GrantId && value == other.value
    public override fun hashCode(): Int = value.hashCode()
}

/** Runtime-owned non-empty receipt identity. */
public class ReceiptId private constructor(public val value: String) {
    public companion object {
        /** Validates one identity. */
        public fun create(value: String): GovernedValueResult<ReceiptId> =
            required(value, "receipt") ?: GovernedValueResult.Success(ReceiptId(value))
    }
    public override fun equals(other: Any?): Boolean = other is ReceiptId && value == other.value
    public override fun hashCode(): Int = value.hashCode()
}

/** Runtime-owned non-empty dispatch-attempt identity. */
public class DispatchAttemptId private constructor(public val value: String) {
    public companion object {
        /** Validates one identity. */
        public fun create(value: String): GovernedValueResult<DispatchAttemptId> =
            required(value, "dispatch attempt") ?: GovernedValueResult.Success(DispatchAttemptId(value))
    }
    public override fun equals(other: Any?): Boolean = other is DispatchAttemptId && value == other.value
    public override fun hashCode(): Int = value.hashCode()
}

/** Exact authority grant for one prepared invocation. */
public data class InvocationGrant(
    public val grantId: GrantId,
    public val invocationId: ToolInvocationId,
    public val preparedDigest: String,
    public val toolName: String,
    public val toolRevision: String,
    public val grantedRequirements: ExecutionRequirements,
    public val constraintsDigest: String,
    public val authorityRevision: String,
)

/** Human/product interaction kind. */
public enum class InteractionKind {
    /** Authority approval. */
    APPROVAL,
    /** Typed external input. */
    EXTERNAL_INPUT,
}

/** Exact interaction request bound to one invocation and digest. */
public data class InteractionRequest(
    public val interactionId: InteractionId,
    public val invocationId: ToolInvocationId,
    public val preparedDigest: String,
    public val kind: InteractionKind,
    public val prompt: JsonElement,
    public val responseSchema: JsonElement,
    public val expiryPolicy: String,
)

/** Typed continuation fact for one interaction. */
public sealed interface InteractionResolution {
    /** Schema-validated durable response. */
    public data class Resolved(
        public val interactionId: InteractionId,
        public val invocationId: ToolInvocationId,
        public val preparedDigest: String,
        public val response: JsonElement,
    ) : InteractionResolution
    /** Durable cancellation. */
    public data class Cancelled(
        public val interactionId: InteractionId,
        public val invocationId: ToolInvocationId,
        public val preparedDigest: String,
    ) : InteractionResolution
}

/** Trustworthy executor terminal classification. */
public enum class TerminalClassification {
    /** Proven success. */
    COMPLETED,
    /** Proven terminal failure. */
    FAILED,
}

/** Trustworthy executor receipt. */
public data class EffectReceipt(
    public val receiptId: ReceiptId,
    public val invocationId: ToolInvocationId,
    public val preparedDigest: String,
    public val grantId: GrantId,
    public val executorId: String,
    public val executorRevision: String,
    public val terminalClassification: TerminalClassification,
    public val resultDigest: String,
)

/** Fact delivered after Runtime durability boundaries complete. */
public sealed interface ExecutionFact {
    /** External dispatch boundary committed. */
    public data class Started(public val dispatchAttemptId: DispatchAttemptId) : ExecutionFact
    /** Successful receipt and bounded content. */
    public data class Completed(
        public val receipt: EffectReceipt?,
        public val content: JsonElement,
        public val truncated: Boolean,
    ) : ExecutionFact
    /** Trustworthy terminal failure. */
    public data class Failed(
        public val receipt: EffectReceipt?,
        public val code: String,
        public val details: String?,
        public val partial: JsonElement?,
    ) : ExecutionFact
    /** Started without trustworthy terminal evidence. */
    public data class Uncertain(public val evidence: String) : ExecutionFact
    /** Requirement cannot be enforced before Started. */
    public data class Unsupported(public val requirement: String) : ExecutionFact
}

/** Model-visible governed observation outcome. */
public sealed interface ObservationOutcome {
    /** Successful bounded tool content. */
    public data class Succeeded(public val content: JsonElement, public val truncated: Boolean) : ObservationOutcome
    /** Policy or interaction rejection. */
    public data class Rejected(public val code: String, public val details: String?) : ObservationOutcome
    /** Trustworthy terminal failure. */
    public data class Failed(public val code: String, public val details: String?, public val partial: JsonElement?) : ObservationOutcome
}

/** Exact model correlation and safe governed outcome. */
public data class GovernedObservation(
    public val invocationId: ToolInvocationId,
    public val preparedDigest: String,
    public val modelCallId: String,
    public val toolName: String,
    public val outcome: ObservationOutcome,
) {
    /** Stable neutral model-visible envelope. */
    public fun modelEnvelope(): JsonObject {
        val fields = linkedMapOf<String, JsonElement>()
        when (val value = outcome) {
            is ObservationOutcome.Succeeded -> fields.putAll(mapOf("status" to JsonPrimitive("succeeded"), "content" to value.content, "truncated" to JsonPrimitive(value.truncated)))
            is ObservationOutcome.Rejected -> { fields.putAll(mapOf("status" to JsonPrimitive("rejected"), "code" to JsonPrimitive(value.code))); value.details?.let { fields["details"] = JsonPrimitive(it) } }
            is ObservationOutcome.Failed -> { fields.putAll(mapOf("status" to JsonPrimitive("failed"), "code" to JsonPrimitive(value.code))); value.details?.let { fields["details"] = JsonPrimitive(it) }; value.partial?.let { fields["partial"] = it } }
        }
        return JsonObject(fields)
    }
}

/** Stable portable reducer failure. */
public enum class GovernedFailureCode {
    /** Grant binding or requirements mismatch. */
    GRANT_MISMATCH,
    /** Executor cannot enforce a requirement. */
    REQUIREMENT_UNSUPPORTED,
    /** Invocation transition conflicts. */
    INVOCATION_CONFLICT,
    /** Interaction binding conflicts. */
    INTERACTION_CONFLICT,
    /** Receipt or recovery state is corrupt. */
    CORRUPT_RECOVERY_STATE,
    /** Model correlation is invalid. */
    INVALID_MODEL_OUTPUT,
}

/** Portable lifecycle state. */
public enum class EffectState { PREPARED, DENIED, REPLACED, AWAITING_INTERACTION, AUTHORIZED, STARTED, COMPLETED, FAILED, UNCERTAIN }

/** Durable recovery position. */
public enum class RecoveryPosition { AUTHORIZED, STARTED_NO_RECEIPT, RECEIPT_NO_RESULT, TERMINAL }

/** Required deterministic recovery action. */
public enum class RecoveryDecision { REVALIDATE_GRANT, RETRY_SAME_INVOCATION, RECOVER_EXECUTOR_RECEIPT, RECONSTRUCT_FROM_RECEIPT, RETURN_TERMINAL, RECONCILE_OPERATOR }

File: engine/tools/src/governed_types.rs
//! Portable C5 authority, interaction, and receipt values.

use std::{error::Error, fmt};

use serde_json::Value;

use crate::ExecutionRequirements;

macro_rules! governed_identity {
    ($name:ident, $label:literal) => {
        #[doc = concat!("Non-empty ", $label, " identity owned by Runtime.")]
        #[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
        pub struct $name(Box<str>);

        impl $name {
            #[doc = concat!("Validates and constructs a ", $label, " identity.")]
            pub fn new(value: impl Into<Box<str>>) -> Result<Self, InvalidGovernanceValue> {
                let value = value.into();
                if value.is_empty() {
                    Err(InvalidGovernanceValue($label))
                } else {
                    Ok(Self(value))
                }
            }

            /// Returns the opaque identity value.
            pub fn as_str(&self) -> &str {
                &self.0
            }
        }
    };
}

governed_identity!(ToolInvocationId, "tool invocation");
governed_identity!(InteractionId, "interaction");
governed_identity!(GrantId, "grant");
governed_identity!(ReceiptId, "receipt");
governed_identity!(DispatchAttemptId, "dispatch attempt");

/// Structural construction failure for a portable C5 value.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct InvalidGovernanceValue(&'static str);

impl fmt::Display for InvalidGovernanceValue {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{} value cannot be empty", self.0)
    }
}

impl Error for InvalidGovernanceValue {}

fn required(value: &str, label: &'static str) -> Result<(), InvalidGovernanceValue> {
    if value.is_empty() {
        Err(InvalidGovernanceValue(label))
    } else {
        Ok(())
    }
}

/// Exact authority grant that may permit one prepared invocation.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct InvocationGrant {
    /// Runtime-owned grant identity.
    pub grant_id: GrantId,
    /// Exact invocation identity.
    pub invocation_id: ToolInvocationId,
    /// Bound C4 digest.
    pub prepared_digest: String,
    /// Bound exact tool name.
    pub tool_name: String,
    /// Bound exact tool revision.
    pub tool_revision: String,
    /// Equal or stricter executor requirements.
    pub granted_requirements: ExecutionRequirements,
    /// Digest of Runtime-owned constraints.
    pub constraints_digest: String,
    /// Exact authority policy revision.
    pub authority_revision: String,
}

impl InvocationGrant {
    /// Rejects empty structural bindings; semantic binding is checked by the reducer.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        grant_id: GrantId,
        invocation_id: ToolInvocationId,
        prepared_digest: impl Into<String>,
        tool_name: impl Into<String>,
        tool_revision: impl Into<String>,
        granted_requirements: ExecutionRequirements,
        constraints_digest: impl Into<String>,
        authority_revision: impl Into<String>,
    ) -> Result<Self, InvalidGovernanceValue> {
        let value = Self {
            grant_id,
            invocation_id,
            prepared_digest: prepared_digest.into(),
            tool_name: tool_name.into(),
            tool_revision: tool_revision.into(),
            granted_requirements,
            constraints_digest: constraints_digest.into(),
            authority_revision: authority_revision.into(),
        };
        required(&value.prepared_digest, "prepared digest")?;
        required(&value.tool_name, "tool name")?;
        required(&value.tool_revision, "tool revision")?;
        required(&value.constraints_digest, "constraints digest")?;
        required(&value.authority_revision, "authority revision")?;
        Ok(value)
    }
}

/// Human/product interaction kind.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum InteractionKind {
    /// Authority approval is required.
    Approval,
    /// Typed external input is required.
    ExternalInput,
}

/// Exact interaction request bound to one invocation and Prepared Call.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct InteractionRequest {
    /// Runtime-owned interaction identity.
    pub interaction_id: InteractionId,
    /// Bound invocation identity.
    pub invocation_id: ToolInvocationId,
    /// Bound C4 digest.
    pub prepared_digest: String,
    /// Required interaction family.
    pub kind: InteractionKind,
    /// Redacted structured prompt.
    pub prompt: Value,
    /// Portable response schema.
    pub response_schema: Value,
    /// Stable Runtime-owned expiry policy reference.
    pub expiry_policy: String,
}

impl InteractionRequest {
    /// Rejects empty bindings and prompts outside the public v1 schema.
    pub fn validate(&self) -> Result<(), InvalidGovernanceValue> {
        required(&self.prepared_digest, "prepared digest")?;
        required(&self.expiry_policy, "expiry policy")?;
        let prompt = self
            .prompt
            .as_object()
            .ok_or(InvalidGovernanceValue("interaction prompt"))?;
        let optional = ["message_text", "cancel_label_key"];
        if prompt.len() < 3
            || prompt.len() > 5
            || prompt.get("schema_version").and_then(Value::as_u64) != Some(1)
            || !non_empty_text(prompt.get("title_key"))
            || !non_empty_text(prompt.get("action_label_key"))
            || optional.iter().any(|key| {
                prompt
                    .get(*key)
                    .is_some_and(|value| !non_empty_text(Some(value)))
            })
            || prompt.keys().any(|key| {
                !matches!(
                    key.as_str(),
                    "schema_version"
                        | "title_key"
                        | "message_text"
                        | "action_label_key"
                        | "cancel_label_key"
                )
            })
        {
            return Err(InvalidGovernanceValue("interaction prompt"));
        }
        Ok(())
    }
}

fn non_empty_text(value: Option<&Value>) -> bool {
    value
        .and_then(Value::as_str)
        .is_some_and(|text| !text.is_empty())
}

/// Typed continuation fact for one requested interaction.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum InteractionResolution {
    /// A schema-validated response was durably committed.
    Resolved {
        /// Exact interaction identity.
        interaction_id: InteractionId,
        /// Exact invocation identity.
        invocation_id: ToolInvocationId,
        /// Exact C4 digest.
        prepared_digest: String,
        /// Typed response retained as durable evidence.
        response: Value,
    },
    /// The interaction was durably cancelled.
    Cancelled {
        /// Exact interaction identity.
        interaction_id: InteractionId,
        /// Exact invocation identity.
        invocation_id: ToolInvocationId,
        /// Exact C4 digest.
        prepared_digest: String,
    },
}

/// Trustworthy executor terminal classification.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TerminalClassification {
    /// Executor proves successful completion.
    Completed,
    /// Executor proves terminal failure.
    Failed,
}

/// Trustworthy receipt binding authority, invocation, executor, and result evidence.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct EffectReceipt {
    /// Runtime-owned receipt identity.
    pub receipt_id: ReceiptId,
    /// Exact invocation identity.
    pub invocation_id: ToolInvocationId,
    /// Exact C4 digest.
    pub prepared_digest: String,
    /// Exact grant identity.
    pub grant_id: GrantId,
    /// Stable executor identity.
    pub executor_id: String,
    /// Exact executor revision.
    pub executor_revision: String,
    /// Proven terminal effect classification.
    pub terminal_classification: TerminalClassification,
    /// Digest of result or terminal evidence.
    pub result_digest: String,
}

impl EffectReceipt {
    /// Rejects empty receipt evidence fields.
    pub fn validate(&self) -> Result<(), InvalidGovernanceValue> {
        required(&self.prepared_digest, "prepared digest")?;
        required(&self.executor_id, "executor identity")?;
        required(&self.executor_revision, "executor revision")?;
        required(&self.result_digest, "result digest")
    }
}
